[jit] Add Custom graph fusion

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#18588 [jit] Add Custom graph fusion**

Differential Revision: [D14901297](https://our.internmc.facebook.com/intern/diff/D14901297)

zdevito reviewed (commented):
I commented on the other WIP TVM IR with my fuser comments. I think they still apply here.

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -1043,7 +1043,7 @@ struct Graph {
       TypePtr typ); // value of None with type Optional[typ]
   TORCH_API Node* createAutogradZero();
-  TORCH_API Node* createFusionGroup();
+  TORCH_API Node* createFusionGroup(Symbol kind = prim::FusionGroup);
```
Since this is no longer specific to `FusionGroup`, maybe it would be better to rename it to `createWithSubgraph(Symbol kind)`?

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -169,10 +169,17 @@ struct GraphFuser {
   Block* block_;
   std::unique_ptr<AliasDb> aliasDb_;
   std::shared_ptr<Graph> graph_;
+  using FusionCallback = std::function<bool(Node*)>;
```
nit: can we move this to not appear inside a sequence of member decalarations? It is quite confusing at the moment

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -169,10 +169,17 @@ struct GraphFuser {
   std::unique_ptr<AliasDb> aliasDb_;
   std::shared_ptr<Graph> graph_;
+  using FusionCallback = std::function<bool(Node*)>;
+  FusionCallback callback_ = [&](Node* n) { return isFusableDefault(n); };
```
Hmm can you really capture `this` in a lambda that's in a default initializer of a member? It's a surprising syntax.

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -1341,5 +1354,11 @@ void FuseGraph(std::shared_ptr<Graph>& graph) {
 }
 
+void CustomFuseGraph(std::shared_ptr<Graph>& graph, std::function<bool(Node*)> fn, Symbol kind) {
+  if (canFuseOnCPU() || canFuseOnGPU()) {
```
My previous comment still applies: why do we check the capabilities of PyTorch fuser, when you're really applying this pass to obtain a fused node that will be passed through a completely different backend.

bwasti committed:
Update on "[jit] Add Custom graph fusion"

[jit] Add Custom graph fusion

gh-metadata: pytorch pytorch 18588 gh/bwasti/2/head

bwasti committed:
Update on "[jit] Add Custom graph fusion"

[jit] Add Custom graph fusion

gh-metadata: pytorch pytorch 18588 gh/bwasti/2/head

wanchaol commented:
clang-tidy is complaining

bwasti committed:
Update on "[jit] Add Custom graph fusion"

[jit] Add Custom graph fusion

gh-metadata: pytorch pytorch 18588 gh/bwasti/2/head

bwasti committed:
Update on "[jit] Add Custom graph fusion"

[jit] Add Custom graph fusion

gh-metadata: pytorch pytorch 18588 gh/bwasti/2/head

bwasti committed:
Update on "[jit] Add Custom graph fusion"

[jit] Add Custom graph fusion

gh-metadata: pytorch pytorch 18588 gh/bwasti/2/head

apaszke commented on torch/csrc/jit/passes/graph_fuser.h:
```diff
@@ -19,6 +19,11 @@ constexpr size_t fusion_kernel_args_limit = 128;
 
+TORCH_API void CustomFuseGraph(
+    std::shared_ptr<Graph>& graph,
+    std::function<bool(Node*)> fn,
```
It would be better to call this `is_fusable` than `fn` because that name carries no information.

apaszke commented on torch/csrc/jit/passes/graph_fuser.h:
```diff
@@ -19,6 +19,11 @@ constexpr size_t fusion_kernel_args_limit = 128;
+TORCH_API void CustomFuseGraph(
+    std::shared_ptr<Graph>& graph,
+    std::function<bool(Node*)> fn,
+    Symbol tag);
```
nit: kind

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -210,7 +225,31 @@ struct GraphFuser {
+  bool isFusableDefault(Node* node) {
+    bool fusableDevice = true;
+    for (const auto& output : node->outputs()) {
+      fusableDevice &= isFusableDevice(output);
```
Please check this only for outputs which have uses.

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -588,21 +614,23 @@ struct GraphFuser {
-        producer->node()->matches(
-            "aten::batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float momentum, float eps, bool cudnn_enabled) -> Tensor")) {
+
+    if (kind_ == prim::FusionGroup &&
```
This place is a bit meh, but I guess it's ok. It's a good sign that we might want to rethink the API you're adding because right now there's no way to do decompose ops lazily with custom fusions.

bddppq commented:
@pytorchbot retest this please

bwasti committed:
Update on "[jit] Add Custom graph fusion"

[jit] Add Custom graph fusion

gh-metadata: pytorch pytorch 18588 gh/bwasti/2/head

bwasti committed:
Update on "[jit] Add Custom graph fusion"

[jit] Add Custom graph fusion

gh-metadata: pytorch pytorch 18588 gh/bwasti/2/head

bwasti committed:
Update on "[jit] Add Custom graph fusion"

[jit] Add Custom graph fusion

gh-metadata: pytorch pytorch 18588 gh/bwasti/2/head

bwasti committed:
Update on "[jit] Add Custom graph fusion"

[jit] Add Custom graph fusion

gh-metadata: pytorch pytorch 18588 gh/bwasti/2/head

bwasti committed:
Update on "[jit] Add Custom graph fusion"

[jit] Add Custom graph fusion

gh-metadata: pytorch pytorch 18588 gh/bwasti/2/head
